### PR TITLE
API updates for Avro encoder

### PR DESCRIPTION
**Pull request by JoshLove-msft**

### CODE REVIEW

**Review by jsquire — APPROVED** [review]

File: sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/README.md [hunk 1]
```diff
@@ -94,11 +94,20 @@ Console.WriteLine(eventData.EventBody);
...
 Console.WriteLine(deserialized.Name);
 ```
 
+### Encode and decode data using `MessageWithMetadata` directly
+
+It is also possible to encode and decode using `MessageWithMetadata`. Use this option if you are not integrating with any of the messaging libraries that work with `MessageWithMetadata`.
+```C# Snippet:SchemaRegistryAvroEncodeDecodeMessageWithMetadata
+var encoder = new SchemaRegistryAvroEncoder(client, groupName, new SchemaRegistryAvroObjectEncoderOptions() { AutoRegisterSchemas = true });
+MessageWithMetadata messageData = await encoder.EncodeMessageDataAsync<MessageWithMetadata>(employee, typeof(Employee));
+
```
Comment (deyaaeldeen) [inline sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/README.md:108]: nit: perhaps print the content of the message so the reader gets familiarized with the shape of the concrete message?

**Review by deyaaeldeen — APPROVED** [review]
Looks good!

File: sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs [hunk 2]
```diff
@@ -217,28 +211,39 @@ private static DatumReader<object> GetReader(Schema writerSchema, Schema readerS
...
+                if (typeof(T) == typeof(MessageWithMetadata))
+                {
+                    // If concrete type is used, we need to use the parameterless constructor as the concrete type does
+                    // not have any other constructors by design (to make it easier to use across the different messaging libraries).
+                    message = (MessageWithMetadata)Activator.CreateInstance(typeof(T));
+                    message.Data = bd;
+                }
+                else
+                {
+                    message = (MessageWithMetadata)Activator.CreateInstance(typeof(T), bd);
```
Comment (christothes) [inline sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs:245]: How do we know the type will implement a ctor accepting `BinaryData`?

Comment (JoshLove-msft) [reply sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs:245]: This is essentially a convention that implementing types will follow. For types that don't adhere to this, a messageFactory needs to be passed.

Comment (JoshLove-msft) [reply sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs:245]: We could possibly improve the error scenario though.

Comment (christothes) [reply sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs:245]: I'm not familiar with this type, so I'm just curious how an implementor would know this? Should the `MessageWithMetadata` implement it too and just set Data to it?

File: sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/README.md (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/README.md:108]: Hmm, the only problem is that the Data is not something that can really be displayed as a string.

File: sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs:245]: In general, other messaging libraries are the implementors - e.g. Event Hubs, Service Bus, and Event Grid.

I was going back and forth on adding the ctor to MessageWithMetadata, but I think you are right - we can just add it. If an inheriting type overrides the property, then the property from the base type will never be used. If they don't override it, then they will have the Data populated.

Comment (JoshLove-msft) [reply sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs:245]: Actually.. I don't think this is correct - adding this constructor to the base type won't help ensure that child types have the same constructor, so I don't think there is a benefit to adding the constructor.

**Pull request merged**